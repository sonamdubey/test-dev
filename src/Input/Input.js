import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import omit from 'omit.js';

function fixControlledValue(value) {
  if (typeof value === 'undefined' || value === null) {
    return '';
  }

  return value;
}

class Input extends React.Component {
  static propTypes = {
    /** A custom class for input container. */
    containerClassName: PropTypes.string,
    /** The default value of the `Input` element. */
    defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** If `true`, the input will be disabled. */
    disabled: PropTypes.bool,
    /** The `id` of the input element. */
    id: PropTypes.string,
    /** A custom class for input. */
    inputClassName: PropTypes.string,
    /** A custom text for label. */
    label: PropTypes.string,
    /** A custom class for label. */
    labelClassName: PropTypes.string,
    /** Name attribute of the `input` element. */
    name: PropTypes.string,
    /** The short hint displayed in the input before the user enters a value. */
    placeholder: PropTypes.string,
    /** The prefix text/icon for the `Input`. */
    prefix: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /** The prefix for component classes. */
    prefixClass: PropTypes.string,
    /** Specify an input field as read-only. */
    readOnly: PropTypes.bool,
    /** Set field as required. */
    required: PropTypes.bool,
    /** The suffix text/icon for the `Input`. */
    suffix: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
    /** Define input type. */
    type: PropTypes.oneOf(['text', 'number', 'tel', 'email', 'password']),
    /** The value of the `input` element. */
    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /** Callback fired when input is blurred. */
    onBlur: PropTypes.func,
    /** Callback fired when the value is changed. */
    onChange: PropTypes.func,
    /** Callback fired when input is focused. */
    onFocus: PropTypes.func,
  };

  static defaultProps = {
    prefixClass: 'oxygen-input',
    type: 'text',
  };

  renderAffixInput(children) {
    const { props } = this;

    if (!('prefix' in props || 'suffix' in props)) {
      return children;
    }

    const prefix = props.prefix && (
      <div className={`${props.prefixClass}-prefix`}>{props.prefix}</div>
    );

    const suffix = props.suffix && (
      <div className={`${props.prefixClass}-suffix`}>{props.suffix}</div>
    );

    return (
      <div className={`${props.prefixClass}-affix-content`}>
        {prefix}
        {children}
        {suffix}
      </div>
    );
  }

  render() {
    const { id, label, value, prefixClass } = this.props;

    let { containerClassName, labelClassName, inputClassName } = this.props;

    containerClassName = classNames([`${prefixClass}`], containerClassName);
    labelClassName = classNames([`${prefixClass}-label`], labelClassName);
    inputClassName = classNames([`${prefixClass}-field`], inputClassName);

    // Fix https://fb.me/react-unknown-prop
    const inputProps = omit(this.props, [
      'containerClassName',
      'labelClassName',
      'inputClassName',
      'label',
      'prefix',
      'prefixClass',
      'suffix',
    ]);

    if ('value' in this.props) {
      inputProps.value = fixControlledValue(value);

      // Input elements must be either controlled or uncontrolled,
      // specify either the value prop, or the defaultValue prop, but not both.
      delete inputProps.defaultValue;
    }

    return (
      <div className={containerClassName}>
        {label && (
          <label htmlFor={id} className={labelClassName}>
            {label}
          </label>
        )}
        {this.renderAffixInput(
          <input {...inputProps} className={inputClassName} />,
        )}
      </div>
    );
  }
}

export default Input;
