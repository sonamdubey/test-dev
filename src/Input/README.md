# Input

## Usage
> Important: Make sure to include the [scss file](https://github.com/carwale/oxygen/blob/master/src/Input/style/input.scss) or feel free to create your own.
```js
import { Input } from "oxygen"
// or
import Input from "oxygen/Input"

<Input
  label="Name"
  type="text"
  placeholder="Enter your name"
  value="John Doe"
  onChange={this.handleChange}
/>
```

## API

### Props

| propName | propType | defaultValue | isRequired | description |
| -------- | -------- | ------------ | ---------- | ----------- |
| containerClassName | `string` | - | - | A custom class for input container. |
| defaultValue | oneOf `string`, `number` | - | - | The default value of the `Input` element. |
| disabled | `bool` | - | - |	If `true`, the input will be disabled. |
| id | `string` | - | - | The `id` of the `input` element. |
| inputClassName | `string` | - | - | A custom class for input. |
| label | `string` | - | - | A custom text for label. |
| labelClassName | `string` | - | - | A custom class for label. |
| name | `string` | - | - | Name attribute of the `input` element. |
| placeholder | `string` | - | - | The short hint displayed in the input before the user enters a value. |
| prefix | oneOf `string`, `node` | - | - | The prefix text/icon for the `Input`. |
| prefixClass | `string` | 'oxygen-input' | - | A prefix for `Input` component classes. |
| readOnly | `bool` | - | - | Specify an input field as read-only. |
| required | `bool` | - | - | Set field as required. |
| suffix | oneOf `string`, `node` | - | - | The suffix text/icon for the `Input`. |
| type | oneOf `text`, `number`, `tel`, `email`, `password` | - | Define input type. |
| value | oneOf `string`, `number` | `''` | - | The value of the `input` element. |


### Events

| propName | propType | defaultValue | isRequired | description |
| -------- | -------- | ------------ | ---------- | ----------- |
| onBlur | `func` | - | - | Callback fired when input is blurred. |
| onChange | `func` | - | - | Callback fired when the value is changed. |
| onFocus | `func` | - | - | Callback fired when input is focused. |
