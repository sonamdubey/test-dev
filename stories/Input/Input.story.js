import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { withKnobs, select, text } from '@storybook/addon-knobs';

import Input from '../../src/Input';
import '../../src/Input/style/input.scss';

/* knobs */
// prop: type
const typeLabel = 'type';
const typeOptions = {
  text: 'text',
  tel: 'tel',
  number: 'number',
  email: 'email',
  password: 'password'
};
const typeDefaultValue = 'text';


storiesOf('Input', module)
  .addDecorator(withKnobs) // TODO: make this a global decoration and move to config file [https://storybook.js.org/addons/introduction/]
  .add('default',
    withInfo()(() => (
      <Input
        id="name"
        label={text('label', 'Name')}
        type={select(typeLabel, typeOptions, typeDefaultValue)}
        placeholder={text('placeholder', 'Enter your name')}
      />
    ))
  )
  .add('with affix',
    withInfo()(() => (
      <Input
        id="mobileNumber"
        label={text('label', 'Mobile Number')}
        type="tel"
        placeholder={text('placeholder', 'Enter your mobile number')}
        value={text('value', '9876543210')}
        maxLength="10"
        prefix="+91"
      />
    ))
  )
