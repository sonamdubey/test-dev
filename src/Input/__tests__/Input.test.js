import React from 'react';
import Input from '../index';

const inputPrefixClassName = 'oxygen-input';

describe('<Input />', function() {
  describe('render', function() {
    it('should render an input', function() {
      const wrapper = mount(
        <Input />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('component props', function() {
    it('should render an input with label', function() {
      const labelText = 'Name';
      const wrapper = mount(
        <Input
          label={labelText}
        />
      );

      const label = wrapper.find('label');
      expect(label).toHaveLength(1);
      expect(label.text()).toEqual(labelText);
      expect(wrapper).toMatchSnapshot();
    })

    it('should render an input with custom classes', function() {
      const labelText = 'Name';
      const containerClassName = 'input-container';
      const labelClassName = 'input-container-label';
      const inputClassName = 'input-container-field';

      const wrapper = mount(
        <Input
          label={labelText}
          containerClassName={containerClassName}
          labelClassName={labelClassName}
          inputClassName={inputClassName}
        />
      );

      expect(wrapper.find(`.${inputPrefixClassName}`).hasClass(containerClassName)).toEqual(true);
      expect(wrapper.find(`.${inputPrefixClassName}-label`).hasClass(labelClassName)).toEqual(true);
      expect(wrapper.find(`.${inputPrefixClassName}-field`).hasClass(inputClassName)).toEqual(true);
    })

    it('should render an input with prefix', function() {
      const mobilePrefix = '+91';
      const wrapper = mount(
        <Input
          type="tel"
          placeholder="Enter mobile number"
          prefix={mobilePrefix}
        />
      )

      const affixContent = wrapper.find(`.${inputPrefixClassName}-affix-content`);
      expect(affixContent).toHaveLength(1);
      const inputPrefix = wrapper.find(`.${inputPrefixClassName}-prefix`);
      expect(inputPrefix).toHaveLength(1);
      expect(inputPrefix.text()).toEqual(mobilePrefix);
      expect(wrapper).toMatchSnapshot();
    })

    it('should render an input with suffix', function() {
      const kilometerSuffix = 'kms';
      const wrapper = mount(
        <Input
          type="tel"
          placeholder="Enter kilometer"
          suffix={kilometerSuffix}
        />
      )

      const affixContent = wrapper.find(`.${inputPrefixClassName}-affix-content`);
      expect(affixContent).toHaveLength(1);
      const inputSuffix = wrapper.find(`.${inputPrefixClassName}-suffix`);
      expect(inputSuffix).toHaveLength(1);
      expect(inputSuffix.text()).toEqual(kilometerSuffix);
      expect(wrapper).toMatchSnapshot();
    })
  });
})
