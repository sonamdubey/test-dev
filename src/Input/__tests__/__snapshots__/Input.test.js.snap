// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> component props should render an input with label 1`] = `
<Input
  label="Name"
  prefixClass="oxygen-input"
  type="text"
>
  <div
    className="oxygen-input"
  >
    <label
      className="oxygen-input-label"
    >
      Name
    </label>
    <input
      className="oxygen-input-field"
      type="text"
    />
  </div>
</Input>
`;

exports[`<Input /> component props should render an input with prefix 1`] = `
<Input
  placeholder="Enter mobile number"
  prefix="+91"
  prefixClass="oxygen-input"
  type="tel"
>
  <div
    className="oxygen-input"
  >
    <div
      className="oxygen-input-affix-content"
    >
      <div
        className="oxygen-input-prefix"
      >
        +91
      </div>
      <input
        className="oxygen-input-field"
        placeholder="Enter mobile number"
        type="tel"
      />
    </div>
  </div>
</Input>
`;

exports[`<Input /> component props should render an input with suffix 1`] = `
<Input
  placeholder="Enter kilometer"
  prefixClass="oxygen-input"
  suffix="kms"
  type="tel"
>
  <div
    className="oxygen-input"
  >
    <div
      className="oxygen-input-affix-content"
    >
      <input
        className="oxygen-input-field"
        placeholder="Enter kilometer"
        type="tel"
      />
      <div
        className="oxygen-input-suffix"
      >
        kms
      </div>
    </div>
  </div>
</Input>
`;

exports[`<Input /> render should render an input 1`] = `
<Input
  prefixClass="oxygen-input"
  type="text"
>
  <div
    className="oxygen-input"
  >
    <input
      className="oxygen-input-field"
      type="text"
    />
  </div>
</Input>
`;
